src/librustc_mir/build/expr/as_rvalue.rs

```diff
@@ -181,6 +181,9 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
             ExprKind::Adt {
                 adt_def, variant_index, substs, fields, base
             } => { // see (*) above
+                let is_union = adt_def.adt_kind() == ty::AdtKind::Union;
+                let active_field_index = if is_union { Some(fields[0].name.index()) } else { None };
+
```

Comment:
> Oh, then the field approach is definitely correct - except for custom Drop impls, which require whole-value initialization "to be armed", what's the plan there?

There is special treatment for types that implement drop, yes. The idea is that you can't have such a type in a "partially initialized" state, so you can't move out from its fields, nor can you write to a field if the type itself is not initialized.